python_modules/dagster/dagster/_core/definitions/metadata/source_code.py

```diff
@@ -130,11 +141,119 @@ def _with_code_source_single_definition(
     return assets_def.with_attributes(metadata_by_key=metadata_by_key)
 
 
+def convert_local_path_to_source_control_path(
+    base_source_control_url: str,
+    repository_root_absolute_path: str,
+    local_path: LocalFileCodeReference,
+) -> UrlCodeReference:
+    source_file_from_repo_root = os.path.relpath(
+        local_path.file_path, repository_root_absolute_path
+    )
+
+    return UrlCodeReference(
+        url=f"{base_source_control_url}/{source_file_from_repo_root}#L{local_path.line_number}",
+        label=local_path.label,
+    )
+
+
+def _convert_local_path_to_source_control_path_single_definition(
+    base_source_control_url: str,
+    repository_root_absolute_path: str,
+    assets_def: Union["AssetsDefinition", "SourceAsset", "CacheableAssetsDefinition"],
+) -> Union["AssetsDefinition", "SourceAsset", "CacheableAssetsDefinition"]:
+    from dagster._core.definitions.assets import AssetsDefinition
+
+    # SourceAsset doesn't have an op definition to point to - cacheable assets
+    # will be supported eventually but are a bit trickier
+    if not isinstance(assets_def, AssetsDefinition):
+        return assets_def
+
+    metadata_by_key = dict(assets_def.metadata_by_key) or {}
```

Comment:
I believe I already tagged @smackesey in at some point where I saw similar code but it would be far preferable in my view to try to structure this to work on a canonicalized `AssetsDefinition`